### PR TITLE
Generate the bulk of the markdown report

### PR DESCRIPTION
**Pull request by kylekthompson**
This PR is part 1 of generating markdown summaries. "part 1" in that it's not yet exposed because there are a few pieces I need to circle back on after a small refactor.

- I need to thread some configuration through
  - Specifically, suite ID, commit sha, branch, captain cloud host, and the substitutions by framework
- Once that refactor is made, I need to:
  - Swap out the real suite ID in place of my TODO
  - Generate retry commands
  - Add a Captain Cloud link if cloud is enabled

Here's one of the snapshots (the more readable one, not the 1MB limit one) rendered: https://gist.github.com/kylekthompson/12058080bdd3fa3a5b4eb03b20eed2c5

### CODE REVIEW

File: internal/reporting/markdown.go [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Skipped, "skipped", "skipped"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Todo, "todo", "todo"); err != nil {
+		return errors.WithStack(err)
+	}
```
Comment (kylekthompson) [inline internal/reporting/markdown.go:164]: We summarize these in the summary line, but we don't have separate sections for them (as they're not all that interesting)

File: internal/reporting/markdown.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+		failedStatus := findFailedStatus(test)
+		markdownTest := markdownTest{
+			Name:     test.Name,
+			Location: location,
+			Retries:  len(test.PastAttempts),
+		}
+		if failedStatus != nil {
+			markdownTest.Message = failedStatus.Message
+			markdownTest.Backtrace = strings.Join(failedStatus.Backtrace, "\n")
+		}
```
Comment (kylekthompson) [inline internal/reporting/markdown.go:317]: Once we have the substitutions and some configuration in here, we can build the retry command and pass it through

File: internal/reporting/markdown.go [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+		testMarkdown := new(strings.Builder)
+		if err := parsedTemplate.Execute(testMarkdown, markdownTest); err != nil {
+			return false, errors.WithStack(err)
+		}
+
+		if oneMB-markdown.Len()-testMarkdown.Len()-len(markdownResultsTruncated) <= 0 {
+			return true, nil
+		}
```
Comment (kylekthompson) [inline internal/reporting/markdown.go:326]: This is where we check if we'll exceed the 1MB limit

File: internal/reporting/markdown_test.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+		Expect(
+			reporting.WriteMarkdownSummary(mockFile, *v1.NewTestResults(v1.RubyRSpecFramework, tests, nil)),
+		).To(Succeed())
+		summary := mockFile.Builder.String()
+		cupaloy.SnapshotT(GinkgoT(), summary)
+
+		Expect(len(summary) < 1000000).To(Equal(true))
+		Expect(strings.Count(summary, hundredKBName.String()) < 10).To(Equal(true))
+	})
```
Comment (kylekthompson) [inline internal/reporting/markdown_test.go:111]: Hopefully this seems comprehensive enough, I tried to do it with smaller tests, but the length of the file was quite large. When I tested manually, I was able to get very close to 1MB, though (~980KB)

File: internal/reporting/.snapshots/Markdown Report produces a truncated summary <= 1MB [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
...aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
Comment (tonywok) [inline internal/reporting/.snapshots/Markdown Report produces a truncated summary <= 1MB:89]: Did you fall asleep on the keyboard again, Kyle?

File: internal/reporting/markdown.go [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+	return false, nil
+}
+
+func pluralize(count int, singular string, plural string) string {
+	if count == 1 {
+		return singular
+	}
+
+	return plural
+}
```
Comment (tonywok) [inline internal/reporting/markdown.go:342]: woo, made it.

**Review by tonywok — APPROVED** [review]
Man, go doesn't make that easy, but this looks to do what you set out to do.

**Pull request merged**